cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/RemoteTaskExecutor.java

```diff
@@ -0,0 +1,100 @@
+import io.cdap.cdap.api.service.worker.RunnableTaskRequest;
+import io.cdap.cdap.common.ServiceUnavailableException;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.http.DefaultHttpRequestConfig;
+import io.cdap.cdap.common.internal.remote.RemoteClient;
+import io.cdap.cdap.common.service.Retries;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.common.service.RetryStrategy;
+import io.cdap.cdap.proto.BasicThrowable;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpRequest;
+import io.cdap.common.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+
+import java.net.NoRouteToHostException;
+import javax.annotation.Nullable;
+
+/**
+ * Helper class for executing a {@link RunnableTaskRequest} on a remote worker.
+ */
+public class RemoteTaskExecutor {
+
+  private static final Gson GSON = new Gson();
+  private static final String TASK_WORKER_URL = "/worker/run";
+  private final RemoteClient remoteClient;
+  private final RetryStrategy retryStrategy;
+
+  public RemoteTaskExecutor(CConfiguration cConf, DiscoveryServiceClient discoveryServiceClient) {
+    this.remoteClient = new RemoteClient(discoveryServiceClient, Constants.Service.TASK_WORKER,
+                                         new DefaultHttpRequestConfig(false),
+                                         Constants.Gateway.INTERNAL_API_VERSION_3);
+    this.retryStrategy = RetryStrategies.fromConfiguration(cConf, Constants.Service.TASK_WORKER + ".");
+  }
+
+  /**
+   * Sends the {@link RunnableTaskRequest} to a remote worker and returns the result.
+   * Retries sending the request if the workers are busy.
+   *
+   * @param runnableTaskRequest {@link RunnableTaskRequest} with details of task
+   * @return byte[] response from remote task
+   * @throws Exception returned by remote task if any
+   */
+  public byte[] runTask(RunnableTaskRequest runnableTaskRequest) throws Exception {
+    HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.POST, TASK_WORKER_URL);
+    HttpRequest httpRequest = requestBuilder.withBody(GSON.toJson(runnableTaskRequest)).build();
+    return Retries.callWithRetries(() -> {
+      try {
+        HttpResponse httpResponse = remoteClient.execute(httpRequest);
+        if (httpResponse.getResponseCode() == HttpResponseStatus.TOO_MANY_REQUESTS.code()) {
+          throw new RetryableException(
+            String.format("Received response code %s for %s", httpResponse.getResponseCode(),
+                          runnableTaskRequest.getClassName()));
+        }
+        if (httpResponse.getResponseCode() != 200) {
+          BasicThrowable basicThrowable = GSON
+            .fromJson(new String(httpResponse.getResponseBody()), BasicThrowable.class);
+          throw createExceptionObject(basicThrowable);
```

Comment:
Since the `createExceptionObject` can return `null`, you should check before throwing. Alternatively, make sure the `createExceptionObject` doesn't return `null`.

Reply:
Changed code to not return null.